ddl/column.go

```diff
@@ -93,57 +99,116 @@ func (d *ddl) createColumnInfo(tblInfo *model.TableInfo, colInfo *model.ColumnIn
 	return colInfo, position, nil
 }
 
-func (d *ddl) onAddColumn(t *meta.Meta, job *model.Job) (ver int64, _ error) {
+func changeAllColumnsState(columnsInfo []*model.ColumnInfo, state model.SchemaState) {
+	for i := range columnsInfo {
+		columnsInfo[i].State = state
+	}
+}
+
+func checkColumnsStateConsistency(columnsInfo []*model.ColumnInfo) bool {
+	state := columnsInfo[0].State
+	for i := 1; i < len(columnsInfo); i++ {
+		if state != columnsInfo[i].State {
+			return false
+		}
+	}
+	return true
+}
+
+func checkHasDuplicateColumnName(columnsInfo []*model.ColumnInfo) error {
+	columnNames := map[string]bool{}
+	for i := 0; i < len(columnsInfo); i++ {
+		name := columnsInfo[i].Name.L
+		if columnNames[name] {
+			return infoschema.ErrColumnExists.GenByArgs(name)
+		}
+		columnNames[name] = true
+	}
+	return nil
+}
+
+// getColumnInfos get all columnInfos correspond to column
+func (d *ddl) getColumnInfos(columns []*model.ColumnInfo, positions []*ast.ColumnPosition, job *model.Job, tblInfo *model.TableInfo,
+	lastColOffset int) ([]*model.ColumnInfo, error) {
+	columnInfos := []*model.ColumnInfo{}
+	var err error
+	for idx, col := range columns {
+		colInfo, _ := findCol(tblInfo.Columns, col.Name.L)
+		pos := &ast.ColumnPosition{Tp: ast.ColumnPositionNone}
+		if idx < len(positions) {
+			pos = positions[idx]
+		}
+
+		if colInfo != nil {
+			if colInfo.State == model.StatePublic {
+				// We already have a column with the same column name.
+				job.State = model.JobCancelled
+				return nil, infoschema.ErrColumnExists.GenByArgs(col.Name)
+			}
```

Comment:
It get the info of column that it in the middle state(StateDeleteOnly->StateWriteOnly->StateWriteReorganization).